File: tool/access_profile_v2_test.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import tempfile
import unittest
from pathlib import Path

import access_profile_v2 as AP

ROOT = Path(__file__).resolve().parents[1]


class AccessProfileV2Test(unittest.TestCase):
    def setUp(self) -> None:
        self.catalog = AP.load_catalog(ROOT / "config/access_profiles.v2.json")

    def test_exact_canonical_profiles_exist(self) -> None:
        self.assertEqual(set(self.catalog), set(AP.PROFILE_IDS))

    def test_round_trip_is_stable(self) -> None:
        for profile in self.catalog.values():
            first = profile.canonical_json()
            second = AP.AccessProfileV2.from_json(profile.to_json()).canonical_json()
            self.assertEqual(first, second, profile.profile_id)

    def test_shared_invalid_vectors_are_rejected(self) -> None:
        fixture = json.loads(
            (ROOT / "evals/fixtures/p1_002_access_profiles/invalid_cases.json").read_text(encoding="utf-8")
        )
        for case in fixture["cases"]:
            with self.subTest(case=case["name"]):
                value = AP.apply_fixture_patch(
                    self.catalog[case["baseProfile"]].to_json(), case["patch"]
                )
                with self.assertRaisesRegex(AP.AccessProfileValidationError, case["errorContains"]):
                    AP.AccessProfileV2.from_json(value)

    def test_catalog_round_trips_from_disk(self) -> None:
        for profile in self.catalog.values():
            with tempfile.TemporaryDirectory() as directory:
                path = Path(directory) / f"{profile.profile_id}.json"
                path.write_text(json.dumps(profile.to_json()), encoding="utf-8")
                self.assertEqual(
                    profile.canonical_json(), AP.AccessProfileV2.from_file(path).canonical_json()
                )

    def test_owner_and_unattended_secret_boundaries(self) -> None:
        owner = self.catalog["owner"].to_json()
        unattended = self.catalog["owner_unattended"].to_json()
        self.assertFalse(owner["sandboxed"])
        self.assertEqual(owner["credentials"]["rawReveal"], "interactive_break_glass")
        self.assertFalse(unattended["sandboxed"])
        self.assertEqual(unattended["credentials"]["rawReveal"], "never")
        self.assertEqual(unattended["process"]["elevation"], "none")


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: test/product/access_profile_v2_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:kristin_local_agent/product/access_profile_v2.dart';

Map<String, dynamic> _copy(Map<String, dynamic> value) {
  return jsonDecode(jsonEncode(value)) as Map<String, dynamic>;
}

void _patch(Map<String, dynamic> target, String dotted, Object? value) {
  final parts = dotted.split('.');
  var current = target;
  for (final part in parts.take(parts.length - 1)) {
    current = (current[part] as Map).cast<String, dynamic>();
  }
  current[parts.last] = value;
}

void main() {
  final root = Directory.current;
  final catalog =
      jsonDecode(
            File(
              '${root.path}/config/access_profiles.v2.json',
            ).readAsStringSync(),
          )
          as Map<String, dynamic>;
  final rawProfiles = (catalog['profiles'] as List).cast<Map>();
  final byId = <String, Map<String, dynamic>>{
    for (final raw in rawProfiles)
      raw['profileId'] as String: raw.cast<String, dynamic>(),
  };

  test('canonical profiles round-trip in Dart', () {
    expect(byId.keys.toSet(), <String>{
      'chat',
      'project',
      'owner',
      'owner_unattended',
      'isolated_untrusted',
    });
    for (final raw in byId.values) {
      final first = AccessProfileV2.fromJson(_copy(raw));
      final second = AccessProfileV2.fromJson(first.toJson());
      expect(
        second.toJson(),
        first.toJson(),
        reason: raw['profileId'] as String,
      );
    }
  });

  test('shared invalid policy vectors fail in Dart', () {
    final fixture =
        jsonDecode(
              File(
                '${root.path}/evals/fixtures/p1_002_access_profiles/invalid_cases.json',
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    for (final rawCase in fixture['cases'] as List) {
      final testCase = (rawCase as Map).cast<String, dynamic>();
      final value = _copy(byId[testCase['baseProfile']]!);
      for (final entry in (testCase['patch'] as Map).entries) {
        _patch(value, entry.key.toString(), entry.value);
      }
      expect(
        () => AccessProfileV2.fromJson(value),
        throwsA(
          isA<AccessProfileValidationException>().having(
            (error) => error.message,
            'message',
            contains(testCase['errorContains'] as String),
          ),
        ),
        reason: testCase['name'] as String,
      );
    }
  });

  test('Owner and Owner unattended remain explicit non-sandbox modes', () {
    final owner = AccessProfileV2.fromJson(_copy(byId['owner']!));
    final unattended = AccessProfileV2.fromJson(
      _copy(byId['owner_unattended']!),
    );
    expect(owner.sandboxed, isFalse);
    expect(owner.credentials['rawReveal'], 'interactive_break_glass');
    expect(unattended.sandboxed, isFalse);
    expect(unattended.credentials['rawReveal'], 'never');
    expect(unattended.process['elevation'], 'none');
  });
}
